### PR TITLE
feat: pivot google sheet

### PR DESCRIPTION
**Pull request by rephus**
Closes:  https://github.com/lightdash/lightdash/issues/12620
Closes: https://github.com/lightdash/lightdash/issues/12730
### Description:
Screenshot from 2024-12-05 15-28-56

on gsheets, we always use raw format 

image

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

File: packages/backend/src/clients/Google/GoogleDriveClient.ts [hunk 1]
```diff
@@ -266,12 +266,6 @@ export class GoogleDriveClient {
             throw new Error('Google Drive is not enabled');
         }
 
-        const auth = await this.getCredentials(refreshToken);
```
Comment (rephus) [inline packages/backend/src/clients/Google/GoogleDriveClient.ts:269]: Moved this gsheet api specific methods to another method that is going to be reused for pivot and not pivot tables

File: packages/backend/src/clients/Google/GoogleDriveClient.ts [hunk 2]
```diff
@@ -306,12 +296,46 @@ export class GoogleDriveClient {
...
 
+        await this.appendCsvToSheet(
+            refreshToken,
+            fileId,
+            [csvHeader, ...values],
+            tabName,
+        );
+    }
+
+    async appendCsvToSheet(
```
Comment (rephus) [inline packages/backend/src/clients/Google/GoogleDriveClient.ts:307]: better name for this methods ?

File: packages/backend/src/scheduler/SchedulerTask.ts [hunk 3]
```diff
@@ -1135,17 +1137,40 @@ export default class SchedulerTask {
...
-                payload.columnOrder,
-                payload.customLabels,
-                payload.hiddenFields,
-            );
+            if (payload.pivotConfig) {
+                // PivotQueryResults expects a formatted ResultRow[] type, so we need to convert it first
+                // TODO: refactor pivotQueryResults to accept a Record<string, any>[] simple row type for performance
+                const formattedRows = formatRows(rows, itemMap);
+
+                const pivotedResults = pivotResultsAsCsv(
```
Comment (rephus) [inline packages/backend/src/scheduler/SchedulerTask.ts:1145]: logic reused and copied from csv service

File: packages/backend/src/services/CsvService/CsvService.ts [hunk 4]
```diff
@@ -621,17 +616,9 @@ This method can be memory intensive
                 metricQuery.customDimensions,
             );
             const customLabels = getCustomLabelsFromTableConfig(config);
-            const hiddenFields = getHiddenTableFields(chart.chartConfig);
-
-            const csvPivotConfig: PivotConfig | undefined = {
```
Comment (rephus) [inline packages/backend/src/services/CsvService/CsvService.ts:626]: extracted conversion into util, and reused in the frontend

File: packages/backend/src/scheduler/SchedulerTask.ts [hunk 5]
```diff
@@ -1135,17 +1137,40 @@ export default class SchedulerTask {
...
-                spreadsheetId,
-                rows,
-                itemMap,
-                payload.showTableNames,
-                undefined, // tabName
-                payload.columnOrder,
-                payload.customLabels,
-                payload.hiddenFields,
-            );
+            if (payload.pivotConfig) {
```
Comment (rephus) [inline packages/backend/src/scheduler/SchedulerTask.ts:1140]: uploadGsheetFromQuery method is used when uploading gsheets from charts , results, or dashboards

File: packages/backend/src/scheduler/SchedulerTask.ts [hunk 6]
```diff
@@ -1607,18 +1633,43 @@ export default class SchedulerTask {
                     undefined,
                     reportUrl,
                 );
+                const pivotConfig = getPivotConfig(chart);
```
Comment (rephus) [inline packages/backend/src/scheduler/SchedulerTask.ts:1636]: this method uploadGsheets is used for scheduler (charts)

File: packages/backend/src/scheduler/SchedulerTask.ts [hunk 7]
```diff
@@ -1721,18 +1772,44 @@ export default class SchedulerTask {
                                 gdriveId,
                                 chartNames[chartUuid] || chartUuid,
                             );
+                        const pivotConfig = getPivotConfig(chart);
```
Comment (rephus) [inline packages/backend/src/scheduler/SchedulerTask.ts:1775]: This is to support `pivot tables` when uploading ghseets on dashboard scheduler. 

We currently don't support `sync on dashboards`, but I updated the code to support this  when we enable the feature.

File: packages/backend/src/clients/Google/GoogleDriveClient.ts [hunk 8]
```diff
@@ -306,12 +296,46 @@ export class GoogleDriveClient {
...
+        const auth = await this.getCredentials(refreshToken);
+        const sheets = google.sheets({ version: 'v4', auth });
+
+        // Clear first sheet before writting
+        await GoogleDriveClient.clearTabName(sheets, fileId, tabName);
+
+        if (results.length === 0) {
+            Logger.info('No data to write to the sheet');
+            return;
+        }
```
Comment (IrakliJani) [inline packages/backend/src/clients/Google/GoogleDriveClient.ts:327]: should this happen before checking credentials and clearing the tab?

File: packages/backend/src/scheduler/SchedulerTask.ts [hunk 9]
```diff
@@ -1141,17 +1144,40 @@ export default class SchedulerTask {
...
+                // TODO: refactor pivotQueryResults to accept a Record<string, any>[] simple row type for performance
+                const formattedRows = formatRows(rows, itemMap);
+
+                const pivotedResults = pivotResultsAsCsv(
+                    payload.pivotConfig,
+                    formattedRows,
+                    itemMap,
+                    payload.metricQuery,
+                    payload.customLabels,
+                    true, // onlyRaw
```
Comment (IrakliJani) [inline packages/backend/src/scheduler/SchedulerTask.ts:1158]: I feel like `pivotResultsAsCsv` should receive an object to make it easy to read what arguments represent

File: packages/common/src/pivotTable/pivotQueryResults.ts [hunk 10]
```diff
@@ -799,7 +800,7 @@ export const pivotResultsAsCsv = (
             );
             const fields = pivotedResults.titleFields[i];
             const fieldLabels = fields.map((field) =>
-                field ? getFieldLabel(field.fieldId) : '-',
+                field ? getFieldLabel(field.fieldId) : undefinedCharacter,
```
[the remark this answers is not in the document]
Comment (IrakliJani) [reply packages/common/src/pivotTable/pivotQueryResults.ts:803]: 👍

**Review by IrakliJani — COMMENTED** [review]
code looks good. added minor suggestions

File: packages/backend/src/clients/Google/GoogleDriveClient.ts (hunk 8 above)
Comment (rephus) [reply packages/backend/src/clients/Google/GoogleDriveClient.ts]: done

File: packages/backend/src/scheduler/SchedulerTask.ts (hunk 9 above)
Comment (rephus) [reply packages/backend/src/scheduler/SchedulerTask.ts]: done

**Review by IrakliJani — APPROVED** [review]

**Pull request merged**